director/projects/api/views/snapshots.py

```diff
@@ -0,0 +1,258 @@
+    lookup_field = "number"
+
+    def get_permissions(self):
+        """
+        Get the list of permissions that the current action requires.
+
+        - `list` and `retrieve`: auth not required but access denied for private projects
+        - `create`: auth required
+        """
+        if self.action == "create":
+            permission_classes = [permissions.IsAuthenticated]
+        else:
+            permission_classes = []
+        return [permission() for permission in permission_classes]
+
+    def get_content_negotiator(self):
+        """
+        Customise the content negotiator for the `retrieve_file` action.
+
+        This avoids DRF raising a not NotAcceptable exception for that
+        action. Instead, the action handles content negotiation itself.
+        """
+        if self.name == "Retrieve file":
+            return IgnoreClientContentNegotiation()
+        else:
+            return super().get_content_negotiator()
+
+    # Views
+
+    def list(self, request: Request, pk: int) -> Response:
+        """
+        List snapshots of the project.
+
+        Returns a list of snapshots created for the project.
+        """
+        project = self.get_project(request.user, pk=pk)
+        if not self.has_permission(ProjectPermissionType.VIEW):
+            raise PermissionDenied
+
+        queryset = project.snapshots.all()
+        serializer = SnapshotSerializer(queryset, many=True)
+        return Response(serializer.data)
+
+    @swagger_auto_schema(
+        request_body=SnapshotCreateRequestSerializer,
+        responses={status.HTTP_201_CREATED: SnapshotSerializer},
+    )
+    def create(self, request: Request, pk: int) -> Response:
+        """
+        Create a snapshot of the project.
+
+        Optionally, receives a `tag` for the new snapshot.
+        Returns the details of the created snapshot.
+        """
+        serializer = SnapshotCreateRequestSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        tag = serializer.validated_data.get("tag")
+
+        # Check that the user has EDIT permissions for the project
+        project = self.get_project(request.user, pk=pk)
+        if not self.has_permission(ProjectPermissionType.EDIT):
+            raise PermissionDenied
+
+        # Create the snapshotter and event for the snapshot
+        snapshotter = ProjectSnapshotter(settings.STENCILA_PROJECT_STORAGE_DIRECTORY)
+        event = ProjectEvent.objects.create(
+            event_type=ProjectEventType.SNAPSHOT.name,
+            project=project,
+            user=logged_in_or_none(request.user),
+            level=ProjectEventLevel.INFORMATIONAL.value,
+        )
+        try:
+            snapshot = snapshotter.snapshot_project(request, project, tag)
+            event.success = True
+            event.save()
+
+            serializer = SnapshotSerializer(snapshot)
+            return Response(serializer.data, status=status.HTTP_201_CREATED,)
```

Comment:
I don't know why linters/black don't complain about this. The trailing `,` is unnecessary.

Reply:
Yes the lack of complaint puzzled me - Black add trailing commas when the function call is multiline (which is why this comma was here is the first place) but is supplosed to remove them if it fits on one line: https://test-black.readthedocs.io/en/style-guide/style_guide/trailing_commas.html